spring-integration-core/src/main/java/org/springframework/integration/endpoint/AbstractPollingEndpoint.java

```diff
@@ -266,7 +267,17 @@ private boolean doPoll() {
 			if (holder != null) {
 				holder.setMessage(message);
 			}
-			this.handleMessage(message);
+			try {
+				this.handleMessage(message);
+			}
+			catch (Exception e) {
+				if (e instanceof MessagingException) {
+					throw new OriginalMessageContainingMessagingException(message, (MessagingException) e);
+				}
+				else {
+					throw new MessagingException(message, e);
+				}
```

Comment:
Looking to this one more time, I'd say it would be much better to have `originalMessage` on the `MessagingException` alongside with the `failedMessage`.
From here, of course, we don't need the same `originalMessage` on the `ErrorMessage` at all...

I think Oleg's idea is like wrap caught `MessagingException` it into a new instance but use `message` as a `failedMessage` already. 

On the client side we would just scan the cause for the targeted `failedMessage`.
That way it would be much confusing, of course, but may work 😉